### PR TITLE
Add support for allUsers(), switchUser(), removeUser()

### PR DESCRIPTION
**Pull request by cmelchior**
Depends on #6783 being merged first.

Adds support for
- `RealmApp.allUsers(): Map<String, RealmUser>`
- `RealmApp.switchUser(user: RealmUser): RealmUser`
- `RealmApp.removeUser(user: RealmUser): RealmUser`
- `RealmApp.removeUser(user: RealmUser, callback: Callback<RealmUser>): RealmUser`

TODO
- [x] Fix remaining tests once #6783 is merged.

### CODE REVIEW

**Comment by cmelchior** [comment]
@nhachicha Ready for review. I just disabled the broken tests right now as they are not related to the PR. There should be issues for all of them.

File: realm/realm-library/src/objectServer/java/io/realm/RealmApp.java [hunk 1]
```diff
@@ -139,12 +142,58 @@ public RealmUser currentUser() {
...
+    }
+
+    /**
+     * Removes a users credentials from this device. If the user was currently logged in, they
+     * will be logged out as part of the process. This is only a local change and does not
+     * affect the user state on the server.
+     *
+     * @param user user to remove.
+     * @param callback callback when removing the user has completed or failed. The callback will always
+     * happen on the same thread as this this method is called on.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/RealmApp.java:186]:
```suggestion
     * happen on the same thread as this method is called on.
```

File: realm/realm-library/src/objectServer/java/io/realm/RealmApp.java [hunk 2]
```diff
@@ -217,7 +270,13 @@ public void logOut() {
      * <p>
      * Once the Realm App has confirmed the logout any registered {@link AuthenticationListener}
      * will be notified and user credentials will be deleted from this device.
+     * <p>
+     * Logging out anonymous users will remove them immediately instead of marking them as
+     * {@link RealmUser.State#LOGGED_OUT}. All other users will be marked as {@link RealmUser.State#LOGGED_OUT}
+     * and will still be returned by {@link #allUsers()}.
      *
+     * @param callback callback when logging out has completed or failed. The callback will always
+     * happen on the same thread as this this method is called on.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/RealmApp.java:279]:
```suggestion
     * happen on the same thread as this method is called on.
```

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/EmailPasswordAuthProviderTests.kt [hunk 3]
```diff
@@ -88,7 +88,7 @@ class EmailPasswordAuthProviderTests {
         val password = "password1234"
         app.emailPasswordAuthProvider.registerUser(email, password)
         val user = app.login(RealmCredentials.emailPassword(email, password))
-        assertEquals(RealmUser.State.ACTIVE, user.state)
+        assertEquals(RealmUser.State.LOGGED_IN, user.state)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/EmailPasswordAuthProviderTests.kt:91]: The names where previously `ACTIVE, ERROR, LOGGED_OUT`, which didn't feel as descriptive, so the where changed to LOGGED_IN, REMOVED and LOGGED_OUT in ObjectStore

File: realm/realm-library/src/main/cpp/io_realm_RealmApp.cpp [hunk 4]
```diff
@@ -104,3 +104,56 @@ JNIEXPORT jobject JNICALL Java_io_realm_RealmApp_nativeCurrentUser(JNIEnv* env,
...
+JNIEXPORT void JNICALL Java_io_realm_RealmApp_nativeRemoveUser(JNIEnv* env,
+                                                               jclass,
+                                                               jlong j_app_ptr,
+                                                               jlong j_user_ptr,
+                                                               jobject j_callback)
+{
+    try {
+        App* app = reinterpret_cast<App*>(j_app_ptr);
+        auto user = *reinterpret_cast<std::shared_ptr<SyncUser>*>(j_user_ptr);
+        app->remove_user(user, JavaNetworkTransport::create_void_callback(env, j_callback));
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/main/cpp/io_realm_RealmApp.cpp:156]: Yes

**Pull request merged**